sklearn/cross_decomposition/kernel_cca.py

```diff
@@ -0,0 +1,449 @@
+        self.coef0 = coef0
+        self.eigen_solver = eigen_solver
+        self.center = center
+        self.pgso = pgso
+        self.eta = eta
+        self.kapa = kapa
+        self.nor = nor
+        self.max_iter = max_iter
+        self.tol = tol
+        self.copy = copy
+
+    def _get_kernel(self, X, Y=None):
+        if callable(self.kernel):
+            params = self.kernel_params or {}
+        else:
+            params = {"gamma": self.gamma,
+                      "degree": self.degree,
+                      "coef0": self.coef0}
+        return pairwise_kernels(X, Y, metric=self.kernel,
+                                filter_params=True, **params)
+
+    def _solve_eigenvalues(self, A):
+        """Solve eigenvalue problem for matrix A"""
+
+        n = A.shape[0]
+
+        if self.eigen_solver == 'auto':
+            if n > 200 and self.n_components < 10:
+                eigen_solver = 'arpack'
+            else:
+                eigen_solver = 'dense'
+        else:
+            eigen_solver = self.eigen_solver
+
+        if eigen_solver == 'dense':
+            rr, alphas = linalg.eigh(
+                A, eigvals=(n - self.n_components, n - 1))
+        elif eigen_solver == 'arpack':
+            rr, alphas = eigsh(A, self.n_components, which="LA",
+                               tol=self.tol, maxiter=self.max_iter)
+        return (rr, alphas)
+
+    def _pgso(self, K, eta):
+        """Partial Gram Schmidt Orthogonalization"""
+
+        m = K.shape[0]
+        index = np.zeros((m, 1))
+        tSize = np.zeros((m, 1))
+        feat = np.zeros((m, m))
+
+        norm2 = np.diag(K)
+        j = 0
+        while np.sum(norm2) > eta and j != m:
+            # find the best new element
+            j2 = np.argmax(norm2)
+
+            # save the index and setting
+            index[j, 0] = j2
+            tSize[j, 0] = np.sqrt(norm2[j2])
+
+            # calculate new features
+            tSum = np.dot(feat[0:m, 0:j], feat[j2, 0:j].T)
+            feat[0:m, j] = (K[0:m, j2] - tSum)/tSize[j]
+
+            # updating diagonal elements
+            norm2 = norm2 - feat[0:m, j]**2
+            j += 1
+        feat = feat[0:m, 0:j]
+        return (feat, tSize, index)
+
+    def _normalisekcca(self, A, K):
+        """Normalise features"""
+
+        n = A.shape[0]
+        KK = np.dot(K.T, K)
+
+        comp = np.sqrt(np.diag(np.dot(np.dot(A.T, KK), A)))
+        comp[comp == 0] = 1
+        comp = comp.reshape((A.shape[1], 1))
+
+        re = A / np.dot(np.ones((n, 1)), comp.T)
+        return re
+
+    def _fit(self, Kx, Ky):
+        """Kernel CCA without PGSO."""
+
+        I = np.diag(np.ones(Kx.shape[0]))
+        Kxinv = linalg.pinv(np.dot((1-self.kapa)*I, Kx) + self.kapa*I)
+        Kyinv = linalg.pinv(np.dot((1-self.kapa)*I, Ky) + self.kapa*I)
+        A = np.dot(np.dot(np.dot(Kxinv, Ky), Kyinv), Kx)
+        A = 0.5*(A.T + A) + np.diag(np.ones(A.shape[0]))*10e-6
+
+        rankA = np.linalg.matrix_rank(A)
```

Comment:
scipy's linalg module may be faster. Can you replace all instances of np.linalg with scipy.linalg?